### DISCUSSION TITLE
Incremental Static Regeneration: disk size

### DISCUSSION THREAD

**Discussion by nicgutierrez** [vercel/next.js#19682] [Help]
Hi 👋 

I'm working on a project that will be used by million of users. Users will be able to access a specific URL that represents their own profiles, i.e `/profile/USER-ID`. The page fetches data from an API and displays that and I'm using ISR to generate those pages only once and then serve the static version for each additional request. 

This sounds like a good approach, at least in theory, but I'm concerned about what this means in terms of generating millions of static html pages on AWS. Is there any recommendation on what's the best approach? Any other performance and optimisation tips are welcome.

Thanks!

**Comment by lfades** [comment]
It entirely depends on whether the profile is **public** or **private**, if it's **private** you **should not** statically generate profile pages with hydrated data, but use CSR instead.

If we're talking about millions of pages, and about data that relies on cookies most likely due to user authentication (private profiles), the best option here is to do CSR.

## Using CSR (Client-Side Rendering)

If `/profile/USER-ID` doesn't use our data fetching methods to get the profile data, and uses client side data fetching to get the user profile, you could render millions of different profiles using the same static page. It will be fast and provide a very reliable experience but it will also require a loading skeleton while the profile data is being fetched.

For example this is how the first load of https://vercel.com/dashboard looks for every registered user:

![image](https://user-images.githubusercontent.com/4278345/100769949-b2bbe200-33ca-11eb-93a0-01c6cbd5716e.png)

That first load then replaces the loading states with the data of the user that requested the page after it has been fetched:

![image](https://user-images.githubusercontent.com/4278345/100770463-48f00800-33cb-11eb-9f6f-3b88d056ed36.png)

The page is static, but the content is dynamically fetched in the browser. Twitter and Youtube are also good examples of this behavior.

## What about SSR?

Also an option, although not that good if the user profile is private, because in that case you won't be able to cache the page and every request will always have to wait for the serverless function to finish, which means that instead of waiting on a skeleton page like in the picture above, you'll be waiting on a server response and users will be more likely to get the feeling that the site is slow.

If the profile is public and doesn't rely on user cookies, you could use [`fallback: 'blocking'`](https://nextjs.org/docs/basic-features/data-fetching#fallback-blocking) so that the first request is `blocking`, meaning the visitor will wait for a serverless function, but any future request will get a static page without having to wait for a serverless function and without a skeleton. In the case of millions of profiles it may only feel faster for profiles that have a lot of views (in which case you could also pre-generate those popular profiles using `getStaticPaths` at build time). An alternative is [`fallback: true`](https://nextjs.org/docs/basic-features/data-fetching#fallback-true) if you prefer a skeleton over a serverless function, but still want to cache the page for future requests.

**Reply by oahmaro** [reply]
Could you suggest what would be better if profile pages are public and not private? Would ISR suites this case or it is better to use SSR? considering a case where i have millions of users
